Add oneof support for Protobuf via Confluent Schema Registry

adamjshook opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

This adds Kafka connector support for Protobuf  `oneOf` data types by encoding them as a JSON object.

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

Protobuf `oneOf` data types are a single field in the message that can hold one value of multiple fields (of varying types) per the schema definition. This adds support for these data types by encoding the entire `oneOf` field as a JSON object where each field name is a key in the JSON object where all values, except one, are `null`. The value that is set corresponds to the set value in the `DynamicMessage` that is processed in the message.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
# Kafka Connector
* Added support for `oneof` Protobuf types when using the Confluent table description provider
```

vlad-lyutenko commented:
I am new for this topic, but I made review, general implementation looks good to me, I have only question about dependencies in pom. A little bit missed with them, why we comment some protobuf plugins?

adamjshook commented:
> I am new for this topic, but I made review, general implementation looks good to me

Thanks for the review, I'm hoping to clean things up today and open this up for review.

>I have only question about dependencies in pom. A little bit missed with them, why we comment some protobuf plugins?

I originally added the protobuf plugins to generate a `TestOneOf` object for unit testing, however by the end of implementation I was able to do everything using a `DynamicMessage`. I commented out the plugins to make sure they weren't needed in any of my tests. The final PR will not use these plugins (I'll be deleting all the commented code).

hashhar commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java:
```diff
@@ -69,43 +79,51 @@ public ProtobufColumnDecoder(DecoderColumnHandle columnHandle)
             this.columnMapping = columnHandle.getMapping();
             this.columnName = columnHandle.getName();
+            this.oneOfs = columnHandle.getOneOfs()
+                    .map(list -> list.stream().sorted(String::compareTo).collect(toImmutableList()));
```
why do we need to sort?

adamjshook replied:
I was sorting it based on [this comment in JsonType](https://github.com/trinodb/trino/blob/f854e103364f67e9c90c1aa3387d1bf00664dda8/core/trino-main/src/main/java/io/trino/type/JsonType.java#L37). Maybe it isn't relevant if the built type is a `Slice` instead of a `Block`.

hashhar replied:
this is very interesting, I had no idea the Trino JsonType required keys to be sorted. I now see in JDBC connectors we don't need to sort because it's done by `JsonTypeUtils#jsonParse`/`JsonTypeUtils#toJsonValue` for us.

wendigo replied:
Maybe we should encode this knowledge?

adamjshook replied:
Since the JSON record we create will only have at most one field in it (since `oneof` can only have one set), sorting the column handles isn't necessary here.

hashhar commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java:
```diff
@@ -118,18 +136,50 @@ private static boolean isSupportedPrimitive(Type type)
             if (valueDescriptor.filter(descriptor -> descriptor.findFieldByName(pathElement) != null).isEmpty()) {
-                return null;
+                if (oneofColumns.isPresent()) {
+                    DynamicSliceOutput output = new DynamicSliceOutput(40);
```
can we come up with better estimates, e.g. based on field count * some avg length?

adamjshook replied:
I made a change to calculate the size based on the string values of the field name, value, plus the extra characters. Looking at the data type test case, there wouldn't be any resizing.

hashhar commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/DecoderColumnHandle.java:
```diff
@@ -33,4 +36,9 @@
 
     String getDataFormat();
+
+    default Optional<List<String>> getOneOfs()
```
nit: Let's consitently use `oneof` instead of `oneOf` since that's what the protobuf schema calls it.

hashhar commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufUtils.java:
separate commit, just formatting changes

adamjshook replied:
I'm just going to revert this; accidentally committed from when I was messing around in this class during implementation.

hashhar commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufValueProvider.java:
```diff
@@ -155,6 +160,14 @@ private static Block serializeObject(BlockBuilder builder, Object value, Type ty
             return serializeRow(builder, value, type, columnName);
         }
 
+        if (type instanceof JsonType) {
+            if (builder != null) {
+                type.writeObject(builder, value);
+                return null;
+            }
+            return (Block) value;
+        }
```
extract to a `serializeJson` method?

hashhar commented on lib/trino-record-decoder/src/test/java/io/trino/decoder/DecoderTestColumnHandle.java:
```diff
@@ -70,6 +72,22 @@
 
+    private final Optional<List<String>> oneOfs;
+
+    public DecoderTestColumnHandle(
```
all ctors should ideally go via the `JsonCreator`, less chance of mistakes

hashhar replied:
if this causes a large diff you can extract as the first commit with some reasons in commit message

hashhar commented on lib/trino-record-decoder/src/test/java/io/trino/decoder/DecoderTestColumnHandle.java:
```diff
@@ -172,7 +199,8 @@ public boolean equals(Object obj)
                 Objects.equals(this.hidden, other.hidden) &&
-                Objects.equals(this.internal, other.internal);
+                Objects.equals(this.internal, other.internal) &&
+                Objects.equals(this.oneOfs, other.oneOfs);
```
list equality depends on element order.

Does a protobuf schema consider differently ordered oneofs as different?
At-least within JSON the field order doesn't matter - so maybe this equality method should do the same?

adamjshook replied:
Can't say the protobuf schema would treat different ordered `oneof` definitions as different objects. The order of the column names in this list matches the order of how the fields are defined in the schema (by name, not by tag number).  Since order doesn't matter, and the final JSON object will have at most one field, how about we change this to a `Set<String>` instead of `List`? @hashhar

hashhar replied:
Set sounds good if we're only going to keep the single field.

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaColumnHandle.java:
```diff
@@ -68,6 +71,22 @@
 
+    private final Optional<List<String>> oneOfs;
+
+    // TODO Should we remove this constructor and update all the tests to use the one annotated with @JsonCreator
```
yes

hashhar replied:
if it's a large diff, extract a commit

hashhar replied:
same applies to other ctor changes as well

hashhar reviewed (commented):
skimmed, mostly editorial comments

Impl looks good.

adamjshook commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/protobuf/TestKafkaProtobufWithSchemaRegistryMinimalFunctionality.java:
```diff
@@ -217,20 +221,48 @@ public void testSchemaWithImportDataTypes()
+
+        assertThat(query(format("SELECT testOneOfColumn FROM %s", toDoubleQuoted(topic))))
+                .matches("""
+                        VALUES (JSON '{"stringColumn":"%s","integerColumn":null,"longColumn":null,"doubleColumn":null,"floatColumn":null,"booleanColumn":null}')
```
Do we want all possible columns in the `oneof` to be in the JSON string, leaving all other values as `null`? Or just a single column e.g. `{"stringColumn":"value"}`?

hashhar replied:
I think having a single field populated would be better because otherwise you can't differentiate that `stringColumn=null` because stringColumn was set to `null` or because some other field existed and stringColumn didn't exist.

adamjshook replied:
Makes sense and I agree. I'll update to only write a field if it is present. Protobuf doesn't do normal `null` values, but I have added a test case for an empty message (which results in an empty JSON value `{}`).

hashhar commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java:
```diff
@@ -33,21 +40,33 @@
 {
+    private static final JsonFactory JSON_FACTORY = new JsonFactoryBuilder().disable(CANONICALIZE_FIELD_NAMES).build();
+    private static final Slice EMPTY_JSON = Slices.utf8Slice("{}");
+    private static final int EXTRA_JSON_CHARACTERS_LENGTH = 7;
```
nit: `"{\"\":\"\"}".length` (or a comment).

hashhar reviewed (commented):
A test showcasing this doesn't work with the file-based table description supplier would be useful.

LGTM from me.

adamjshook commented:
>A test showcasing this doesn't work with the file-based table description supplier would be useful.

Added a test case asserting the column names are not found when using the file-based descriptor.

hashhar commented:
@mosabua / @jhlodin can you please review the last commit which includes documentation changes.

Praveen2112 commented:
> Flipping this back to review for merge in the current state. We can follow up on any improvements later on.
Can we have a GH issue to track the followup ?

Praveen2112 commented on plugin/trino-kafka/src/test/resources/protobuf/test_oneof.proto:
```diff
@@ -0,0 +1,12 @@
+        uint64 longColumn = 3;
+        double doubleColumn = 4;
+        float floatColumn = 5;
+        bool booleanColumn = 6;
```
Can we also pass a `message` type and check on how `DynamicMessage` would be serialized.

Praveen2112 commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java:
```diff
@@ -144,4 +177,36 @@ private static Optional<Descriptor> getDescriptor(FieldDescriptor fieldDescripto
         return Optional.empty();
     }
+
+    private static Object createOneofJson(DynamicMessage message, Set<String> oneofColumns)
```
Any specific reason on why we should pass the `oneofColumns` from the KafkaColumnHandle ? Why can't we use `message.getDescriptorForType()` get oneOf types from here and handle them ? In this case we don't have to change other decoder right ?

wendigo replied:
Yeah. You can collect `oneof` fields from `FieldDescriptor`:

`.getContainingOneof().getName()`

adamjshook replied:
Great, thanks for the suggestion. I just couldn't find a way to get the names to where they needed to be, but this does the trick. This removes the need to include the `oneof` names in the column handles.

Praveen2112 commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufValueProvider.java:
```diff
@@ -139,11 +144,15 @@ private static Slice getSlice(Object value, Type type, String columnName)
 
     @Nullable
-    private static Block serializeObject(BlockBuilder builder, Object value, Type type, String columnName)
+    private Block serializeObject(BlockBuilder builder, Object value, Type type, String columnName)
```
nit: These can be extracted to a new commit.

Praveen2112 commented on lib/trino-record-decoder/src/test/java/io/trino/decoder/protobuf/TestProtobufDecoder.java:
```diff
@@ -104,12 +116,136 @@ public void testAllDataTypes(String stringData, Integer integerData, Long longDa
+                "booleanColumn",
+                "numberColumn",
+                "timestampColumn",
+                "bytesColumn");
```
Can we add additional columnName with `message` type

Praveen2112 commented on lib/trino-record-decoder/src/test/java/io/trino/decoder/protobuf/TestProtobufDecoder.java:
```diff
@@ -104,12 +116,136 @@ public void testAllDataTypes(String stringData, Integer integerData, Long longDa
+                    .build();
+            assertOneof(message, testOneOfColumn, oneofColumnNames, Map.of("booleanColumn", booleanData));
+        }
+    }
```
for timestamp ?

Praveen2112 reviewed (commented):
Overall looks good to me. 


But I would like to see how it serializes/deserializes if one of `oneOf` field is of a `message` type - From this docs - https://protobuf.dev/programming-guides/proto3/#using-oneof it looks like `map` and `array` are not supported but `row` should be supported by `oneOf` 

And if it is possible for us to determine the oneOf fields when reading the message, do we need to propagate these information when we create `KafkaColumnHandle` ?

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
 insensitive match must be done, as identifiers cannot contain upper case
 characters.
 
+Protobuf specific type handling in Confluent table description supplier
```
```suggestion
Protobuf-specific type handling in Confluent table description supplier
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+Protobuf specific type handling in Confluent table description supplier
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+When using the Confluent table description supplier the following Protobuf
```
```suggestion
When using the Confluent table description supplier, the following Protobuf
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+When using the Confluent table description supplier the following Protobuf
+specific types are supported in addition to the :ref:`normally supported
```
```suggestion
specific types are supported in addition to the :ref:`normally supported types
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+
+When using the Confluent table description supplier the following Protobuf
+specific types are supported in addition to the :ref:`normally supported
+<kafka-protobuf-decoding>` types:
```
```suggestion
<kafka-protobuf-decoding>`:
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+oneof
++++++
+
+Protobuf schemas containing ``oneof`` fields get mapped to a ``JSON`` field in
```
```suggestion
Protobuf schemas containing ``oneof`` fields are mapped to a ``JSON`` field in
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+Protobuf schemas containing ``oneof`` fields get mapped to a ``JSON`` field in
+Trino.
+
+For example given the following Protobuf schema:
```
```suggestion
For example given the following Protobuf schema:
```
```suggestion
For example, given the following Protobuf schema:
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+        }
+    }
+
+The corresponding Trino row would have a ``JSON`` field ``test_oneof_column``
```
```suggestion
The corresponding Trino row is a ``JSON`` field ``test_oneof_column``
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+containing a JSON object with a single key. The value of the key would match
+the name of the ``oneof`` type which was actually present.
+
+In the above example it means if the Protobuf message had the
```
```suggestion
In the above example, if the Protobuf message has the
```

mosabua commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+
+In the above example it means if the Protobuf message had the
+``test_oneof_column`` containing ``string_column`` set to a value ``Trino``
+then the corresponding Trino row would have a column named
```
```suggestion
then the corresponding Trino row includes a column named
```

mosabua reviewed (commented):
A couple of fixes for docs necessary. Good once the suggestions are applied.

jhlodin commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -479,6 +480,45 @@ used to resolve the subject name via the topic name. Note that a case
+
+The corresponding Trino row would have a ``JSON`` field ``test_oneof_column``
+containing a JSON object with a single key. The value of the key would match
+the name of the ``oneof`` type which was actually present.
```
```suggestion
the name of the ``oneof`` type that is present.
```

jhlodin reviewed (commented):
One additional comment, +1 to Manfred that docs are good once changes are applied

wendigo commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/DecoderColumnHandle.java:
```diff
@@ -33,4 +36,9 @@
 
     String getDataFormat();
+
+    default Optional<Set<String>> getOneofs()
```
nit: `getOneOfs` ?

adamjshook replied:
We've been styling as as `oneof`/`Oneof` to match how Protobuf styles it.

wendigo replied:
Makes sense!

wendigo commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/protobuf/ProtobufColumnDecoder.java:
```diff
@@ -144,4 +177,36 @@ private static Optional<Descriptor> getDescriptor(FieldDescriptor fieldDescripto
+
+    private static Object createOneofJson(DynamicMessage message, Set<String> oneofColumns)
+    {
+        // Find the oneof field in the message; there will be at most one
```
Is the "at most one" limitation of the protobuf?

wendigo replied:
https://protobuf.dev/programming-guides/proto3/#using-oneof as far as I can tell, there could be more then just one named `oneof`s in the protobuf message.

wendigo replied:
We should rather than pass the `oneofColumns` enumerate those using `DynamicMessage`

adamjshook replied:
You can specify more than one `oneof` fields in the message schema, but the field names across all `oneof` types have to be distinct so you'll only find one.

I've made the change to remove using `oneofColumns` in favor of enumerating the field names in `DynamicMessage`.

hashhar replied:
> but the field names across all oneof types have to be distinct so you'll only find one.

Is this verified by the protobuf schema parser or something else? AFAIK the protobuf schema is based on tags so the schema itself probably doesn't care.

adamjshook replied:
Given a schema file:

```
syntax = "proto3";
message schema {
    oneof testOneofColumn {
        string stringColumn = 1;
    }
    string stringColumn = 2;
}
```

Loading the file-based schema parser doesn't result in any failure, but there is only one `stringColumn` field defined in the `Descriptor` (I'd say because it doesn't handle `oneof` types in the first place, so it is ignored).

The Confluent parser doesn't result in any failures, but there is only one `stringColumn` field in the resulting descriptor (it seems to flatten all of the `oneof` fields into a single field description, so having an additional field with the same name will be de-duplicated.

Meanwhile, `protoc` fails (specifically the Maven plugin which runs `protoc`) fails but this may just be an artifact of the code generation:

```
protoc-jar: executing: [/tmp/protocjar14095071959677501612/bin/protoc.exe, -I/tmp/protocjar6731110910677765846/include, -I/home/adamjshook/gitrepos/trino/lib/trino-record-decoder/src/test/resources/decoder/protobuf, --java_out=/home/adamjshook/gitrepos/trino/lib/trino-record-decoder/target/generated-test-sources, /home/adamjshook/gitrepos/trino/lib/trino-record-decoder/src/test/resources/decoder/protobuf/test_oneof.proto]
test_oneof.proto:7:12: "stringColumn" is already defined in "schema".
```

hashhar replied:
Yeah, seems protobuf is under-specified for this case and behaviour depends on impl. Since messages will be produced using Confluent serializer i think the assumption holds true.

hashhar replied:
Can we make it throw if we find multiple?

wendigo commented on lib/trino-record-decoder/src/test/java/io/trino/decoder/DecoderTestColumnHandle.java:
```diff
@@ -91,6 +96,7 @@ public DecoderTestColumnHandle(
         this.keyDecoder = keyDecoder;
         this.hidden = hidden;
         this.internal = internal;
+        this.oneofs = oneofs;
```
add `requireNonNull` - oneofs is not a primitive field

adamjshook commented:
Pushed after applying the feedback here. Noteable changes are:

- Removed having the `oneofs` on the `DecoderColumnHandle` and the `KafkaTopicFieldDescription`. They are now sourced from the message descriptor in the `ProtobufColumnDecoder`
- Added the `Row` and `NestedRow` data types to the `test_oneof.proto` file
- Changed how the JSON value is being built in `ProtobufColumnDecoder` to use the `JsonFormat.Printer` from the `protobuf-java-util`

hashhar commented:
just rebased on master for fresh ci results, will merge soon
